src/hotspot/share/gc/shared/cardTable.hpp

```
@@ -199,7 +200,9 @@ class CardTable: public CHeapObj<mtGC> {
   // This would be the 0th element of _byte_map, if the heap started at 0x0.
   // But since the heap starts at some higher address, this points to somewhere
   // before the beginning of the actual _byte_map.
-  CardValue* byte_map_base() const { return _byte_map_base; }
+  virtual CardValue* byte_map_base() const { return _byte_map_base; }
+
+  virtual CardValue* byte_map() const { return _byte_map; }
```

@shipilev - can you please confirm that this is the part that you didn't like?

---

Yes, I am not fond of extending `CardTable` with virtual members, especially if they can be used on high-performance paths. Not sure if the following idea is viable. 

ShenandoahBarrierSet knows where to get card table base: from Shenandoah thread local data. Now it looks like we need to deal with two problems:
  1. Protect ourselves from accidentally calling `CardTable` methods that may reference "incorrect" `_byte_map_(base)`. To do that, it looks it is enough to initialize `CardTable::_byte_map_(base)` to non-sensical values (`nullptr`-s?), and let the testing crash.
  2. Allow calls to `CardTable` utility methods with our base. For that, I think we can drill a few new (non-virtual) methods in `CardTable`, and enter from Shenandoah through them. So for example `byte_for_index(const size_t card_index)` becomes:
  ```
  CardValue* byte_for_index(const CardValue* base, const size_t card_index) const {
    return base + card_index;
  }
  CardValue* byte_for_index(const size_t card_index) const {
    return byte_for_index(_byte_map, card_index);
  }
  ```

---

@shipilev - can you please take a look at the latest pushes? I realized that the logic implemented already keeps the fields of the base card table class always updated, therefore I don't really need to make the methods (`_byte_map_(base)` virtual at all.